### DISCUSSION TITLE
How can I implement a Parser that never accepts input?

### DISCUSSION THREAD

**Discussion by nathan-at-least** [zesterer/chumsky#139] [Q&A] [answered]
I'm writing a recursive parser for a generic type which can be specialized to subgrammars, something like:

```
enum Expr<T> {
    Deref(String),
    Add(Box<Expr<T>>, Box<Expr<T>>),
    Extension(T),
}
```

One kind of `T` is effectively `Infallable`, which handles the case of the simplest common grammar. Other cases might extend the grammar by creating subparsers specific to `T`.

For the `Infallable` case, how can I write a parser that never accepts input?

A sketch of my current approach for `Expr` might look something like this:

```
trait Parseable {
    type P: Parser<char, Self, Error=Simple<char>>;

    fn parser() -> P;
}

fn expr<T>(expr: Recursive<...>) -> impl Parser<char, Expr<T>, Error=Simple<char>>
    where T: Parseable,
{
    ident().map(Expr::Deref)
      .or(parse_add_expr(expr))
      .or(T::parser()).map(Expr::Extension)
}
```

So my question is how to define an `impl Parseable for Infallible` (or some new-type which means "no extension syntax").

Also, aside from my main question, if anyone has advice for how to do this "extensible subgrammar" design, I'm interested in that, too. I haven't yet figured out if this is feasible with `recursive`.

**Comment by nathan-at-least** [comment]
Ok, I just found [chumsky](https://docs.rs/chumsky/0.8.0/chumsky/index.html)::[primitive](https://docs.rs/chumsky/0.8.0/chumsky/primitive/index.html)::[empty](https://docs.rs/chumsky/0.8.0/chumsky/primitive/fn.empty.html#), which may work.

**Reply by nathan-at-least** [reply]
Whoops, that is not what I want. `empty` parses nothing and succeeds with `()`, yet I need something that parses nothing and fails. Let's say it's called `nothing`, then `foo.or(nothing())` should behave exactly as `foo`.

I'll take a stab at writing `nothing`.

**Comment by nathan-at-least** [comment]
It seems that this has the right behavior except it may produce confusing error messages:

```rust
empty().validate(|(), span, emit| emit(Simple::custom(span, "<internal non-parsing case>")))
```

**Reply by nathan-at-least** [reply]
The challenge with this approach is that I'm implementing it for an empty enum (similar to `Infallible`). Let's call it `enum NullSyntax {}`, so this parser must implement `Parser<_, NullSyntax>`. The type of the parser in this example is `Parser<_, ()>`. I'm wondering why the `validate` callback takes `emit` and returns `U` rather than returning `Result<U, Error>`?

**Reply by zesterer** [reply]
It seems that `try_map` is probably the right combinator. That said, I think there are good use-cases for a `failure()` parser that always generates an error, and perhaps another named `success()` for the opposing behaviour.

**Reply by nathan-at-least** [reply]
Yep, `try_map` works like a charm. Kind of embarrassing how far down the wrong rabbit-hole I went, but hopefully someone else might learn from this discussion. ;-)

**Comment by nathan-at-least** [comment, marked as answer]
As pointed out in https://github.com/zesterer/chumsky/discussions/139#discussioncomment-2753293, I can achieve this simply with [`chumsky::Parser::try_map`](https://docs.rs/chumsky/0.8.0/chumsky/trait.Parser.html#method.try_map):

```rust
empty().try_map(|(), span| Err(Simple::custom(span, "<unparsable internal case>")))
```
